### PR TITLE
Added VFS support

### PR DESCRIPTION
**Pull request by albertofustinoni**
Core loads BIOS from files. Added support to do so via libretro VFS

### CODE REVIEW

File: libretro/Makefile.common [hunk 1]
```diff
@@ -25,3 +25,11 @@ SOURCES_C := \
...
 	$(CORE_DIR)/src/ymfm/ymfm_ssg.c \
 	$(CORE_DIR)/src/z80/z80.c
+    $(CORE_DIR)/libretro/libretro-common/compat/compat_posix_string.c
+    $(CORE_DIR)/libretro/libretro-common/compat/compat_strl.c
+    $(CORE_DIR)/libretro/libretro-common/compat/fopen_utf8.c
+    $(CORE_DIR)/libretro/libretro-common/encodings/encoding_utf.c
+    $(CORE_DIR)/libretro/libretro-common/streams/file_stream.c
+    $(CORE_DIR)/libretro/libretro-common/streams/file_stream_transforms.c
+    $(CORE_DIR)/libretro/libretro-common/string/stdstring.c
+    $(CORE_DIR)/libretro/libretro-common/vfs/vfs_implementation.c
```
Comment (carmiker) [inline libretro/Makefile.common:35]: Inconsistent indentation, and you will also need to add a \ to everything but the last file in the list.

File: src/geo.c [hunk 2]
```diff
@@ -43,6 +43,10 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #include "geo_ymfm.h"
 #include "geo_z80.h"
 
+#ifdef __LIBRETRO__
```
Comment (carmiker) [inline src/geo.c:46]: Why is this necessary?

File: src/geo.c [hunk 3]
```diff
@@ -241,6 +245,7 @@ int geo_bios_load_mem(void *data, size_t size) {
     return geo_bios_load(&zip_archive);
 }
 
+#ifndef __LIBRETRO__
```
Comment (carmiker) [inline src/geo.c:248]: Is there a reason this needs to be left out of the build?
Please refer to the "Contributing" section of the README. This sort of thing should be avoided in favour of generic solutions.

File: src/geo.c (hunk 2 above)
Comment (albertofustinoni) [reply src/geo.c:46]: Because adding `file_stream_transforms.h` causes `stdio` functions like fopen to be mapped to retroarch's VFS implementation. If you are building the core standalone this would cause the build to fail.

File: src/geo.c (hunk 3 above)
Comment (albertofustinoni) [reply src/geo.c:248]: I do not know how the core is meant to work standalone, outside of libretro: if you don't have access to the VFS functionality you may as well continue to keep using `miniz`'s functionmality to load archives from files directly instead of buffering to memory through the VFS

File: libretro/Makefile.common (hunk 1 above)
Comment (albertofustinoni) [reply libretro/Makefile.common]: Fixed

File: src/geo.c (hunk 3 above)
Comment (carmiker) [reply src/geo.c:248]: In this case, let's just leave it how it is now. There shouldn't be any harm in keeping this function in the build.

File: src/geo.c (hunk 2 above)
Comment (carmiker) [reply src/geo.c:46]: This is not something I can get on board with. Basically, there should be no #ifdefs added for any specific platform anywhere in the `src/` directory. If there is a case where something inside here breaks, then it's a case where a generic implementation can replace whatever breaks, and any platform-specific code can piggyback on that from the outside.

If you take a look at `geo_mem_ptr` it gives you the ability to manipulate any raw memory block in the emulator. This can be used to build libretro VFS implementations in libretro.c. I believe the only places this would be necessary would be for memory cards, NVRAM, and (rarely, just jockeygp and vliner) cartridge RAM aka SRAM. If this is not enough, then I can expose whatever else is necessary to do everything in libretro.c.

**Comment by albertofustinoni** [comment]
Reverted the changes in `geo.c`

Would it be possible to implement `retro_get_memory_data` for `RETRO_MEMORY_SAVE_RAM`?

With the changes in `libretro.c` the core can get ingame (rom passed as buffer and BIOS loaded via VFS), but as is there's no way around `stdio` save game data.

**Comment by carmiker** [comment]
> Reverted the changes in `geo.c`
> 
> Would it be possible to implement `retro_get_memory_data` for `RETRO_MEMORY_SAVE_RAM`?
> 
> With the changes in `libretro.c` the core can get ingame (rom passed as buffer and BIOS loaded via VFS), but as is there's no way around `stdio` save game data.

This is what I tried when I started writing the port, but it gets complicated because there are 3 different types of save data. Arguably the NVRAM could fall under RETRO_MEMORY_RTC as it does contain RTC data, but it also includes a lot more than that. Actual SRAM is only used by two titles, the gambling titles from Brezzasoft (vliner and jockeygp), and the main purpose is to make sure if the power goes out, you don't lose your score. Then there are memory cards. I think trying to work this int `retro_get_memory_data` is going to just result in ugly hacks.

Probably the best way to do it is to create something based on `geo_mem_ptr`. What I am envisioning is 2 possible codepaths: the current one, and the VFS one, chosen in libretro.c based on whether the frontend has a VFS implementation or not. This way it continues to work as it does now for simpler frontends, but for RA it always uses VFS.

@albertofustinoni if you are okay with it, I can take this to the finish line from where it is now, but I don't want to step on your toes if you'd like to do it.

**Comment by carmiker** [comment]
@albertofustinoni I have rewritten the implementation in this PR, but it still lacks VFS support for save data. I'll do that a bit later. Thanks for getting the ball rolling here.
